### PR TITLE
新增1a2b小遊戲

### PR DESCRIPTION
**Pull request by wow890209**
## Why need this change? / Root cause: 
-新增1a2b小遊戲
## Changes made:
-
## Test Scope / Change impact:
-

### CODE REVIEW

File: guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    var gameRooms = mutableListOf<GameRoom>()
+
+    fun register(member: Member, channel: ThreadChannel) {
+        gameRooms.add(GameRoom(member, channel))
+        logger.info {
+            "[GameRepository register event] " + "{" + "\"feature\" : \"guess-num-1a2b\"," + "\"game room count\" : \"${gameRooms.size}\"," + "}"
+        }
+    }
+
+    fun unregister(member: Member, channel: ThreadChannel) {
```
Comment (Wally5077) [inline guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt:35]: ditto **Discord Model 轉成自己的 Model**

File: guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+        val deleteRoom = gameRooms.filter { room ->
+            room.member == member && room.threadChannel == channel
+        }.onEach { it.close() }
+        gameRooms -= deleteRoom
+
+        logger.info {
+            "[GameRepository unregister event] " + "{" + "\"feature\" : \"guess-num-1a2b\"," + "\"game room count\" : \"${gameRooms.size}\"," + "}"
```
Comment (Wally5077) [inline guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt:43]:
```kotlin
"[GameRoom Unregistered]  { \"gameRoomCount\" : \"${gameRooms.size}\"}"
```

File: guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    fun register(member: Member, channel: ThreadChannel) {
+        gameRooms.add(GameRoom(member, channel))
+        logger.info {
+            "[GameRepository register event] " + "{" + "\"feature\" : \"guess-num-1a2b\"," + "\"game room count\" : \"${gameRooms.size}\"," + "}"
+        }
+    }
+
+    fun unregister(member: Member, channel: ThreadChannel) {
+
+        val deleteRoom = gameRooms.filter { room ->
```
Comment (Wally5077) [inline guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt:37]: deleteRoom -> unregisteredGameRoom

File: guess-num-1A2B/src/test/kotlin/domainTest.kt [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        assertEquals(game.guess("7890"), "0A0B")
+        assertEquals(game.guess("123"), "0A0B")
+        assertEquals(game.guess("12345"), "0A0B")
+    }
+
+    @Test
+    fun `guess the number`() {
+        assertEquals(game.guess("4321"), "0A4B")
+    }
+}
```
Comment (Wally5077) [inline guess-num-1A2B/src/test/kotlin/domainTest.kt:43]: 這是 ChatGPT 產的
```kotlin
class GuessNum1A2BTest {
    private val secretNumber = "1234"
    private val game = GuessNum1A2B(secretNumber)

    @Test
    fun `test when guess has only As`() {
        assertGuessResult("1567", "1A0B")
        assertGuessResult("1267", "2A0B")
        assertGuessResult("1236", "3A0B")
        assertGuessResult("1234", "4A0B")
    }

    @Test
    fun `test when guess has only Bs`() {
        assertGuessResult("5671", "0A1B")
        assertGuessResult("5621", "0A2B")
        assertGuessResult("5321", "0A3B")
    }

    @Test
    fun `test when guess has both As and Bs`() {
        assertGuessResult("1423", "1A3B")
        assertGuessResult("1243", "2A2B")
    }

    @Test
    fun `test when guess has neither As nor Bs`() {
        assertGuessResult("7890", "0A0B")
        assertGuessResult("123", "0A0B")
        assertGuessResult("12345", "0A0B")
    }

    @Test
    fun `test when guess is the same as the secret number`() {
        assertGuessResult(secretNumber, "4A0B")
    }

    @Test
    fun `test with invalid guess input`() {
        assertThrows<IllegalArgumentException> { game.guess("12a4") }
        assertThrows<IllegalArgumentException> { game.guess("12345") }
        assertThrows<IllegalArgumentException> { game.guess("12") }
        assertThrows<IllegalArgumentException> { game.guess("") }
    }

    private fun assertGuessResult(guess: String, expected: String) {
        val result = try {
            game.guess(guess)
        } catch (ex: IllegalArgumentException) {
            throw AssertionError("Input validation failed for guess $guess", ex)
        }
        assertEquals(expected, result)
    }
}
```

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done.

File: guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+import tw.waterballsa.utopia.guessnum1a2b.domain.generateSecretNumber
+import tw.waterballsa.utopia.jda.listener
+import java.util.*
+import java.util.concurrent.TimeUnit
+import kotlin.concurrent.timerTask
+import kotlin.time.Duration.Companion.minutes
+
+val logger = KotlinLogging.logger {}
+val timer = Timer()
+val repository = GameRepository()
```
Comment (frankvicky) [inline guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt:23]: private

File: guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    fun clear() {
+        gameRooms.forEach { room ->
+            room.close()
+        }
+        gameRooms.clear()
+
+    }
+}
+
+class GameRoom(val member: Member, val threadChannel: ThreadChannel) {
```
Comment (frankvicky) [inline guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt:69]: `(private  val member: Member, private val threadChannel: ThreadChannel)`

File: guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        }, 10.minutes.inWholeMilliseconds)
+    }
+
+    fun guess(number: String): String {
+        if (validateMessage(number)) {
+            return "你的答案不是4個數字"
+        }
+        if (isEnd) {
+            return "遊戲已結束"
+        }
```
Comment (frankvicky) [inline guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt:86]: when

File: guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+        var result = game.guess(number)
+        if (result == correctAnswer) {
+            result += "恭喜猜對!"
```
Comment (frankvicky) [inline guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt:90]: String interpolate

File: guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        return result
+    }
+
+    fun close() {
+        threadChannel.sendMessage("遊戲房間將於10秒後關閉").queueAfter(1, TimeUnit.SECONDS)
+        threadChannel.retrieveParentMessage().queueAfter(10, TimeUnit.SECONDS) { startMessage ->
+            startMessage.delete().queue()
+            threadChannel.delete().queue()
+        }
+    }
```
Comment (frankvicky) [inline guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt:103]:
```kotlin
threadChannel.run {
    // action
}
```

File: guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+                reply("1A2B遊戲開始! 將於10分鐘後關閉，$asMention 遊玩愉快~").queue {
+                    hook.retrieveOriginal().queue { message ->
+                        message.createThreadChannel("$effectiveName's room").queue { threadChannel ->
+                            repository.register(this, threadChannel)
+                        }
+                    }
+                }
+            } else {
+                reply("你已經有一間房間了").setEphemeral(true).queue()
+            }
```
Comment (frankvicky) [inline guess-num-1A2B/src/main/kotlin/tw/waterballsa/utopia/guessnum1a2b/listener.kt:130]:
```kotlin
repository.find(member!!)?.let { 
    //action 
}
```

**Review by frankvicky — CHANGES_REQUESTED** [review]

**Pull request closed without merging**